### PR TITLE
Add classes for memory dumps

### PR DESCRIPTION
**Pull request by petrcermak**
This patch adds two classes, <tt>WorldMemoryDump</tt> and <tt>ProcessMemoryDump</tt>, which will be used to model memory dumps of processes.

### CODE REVIEW

File: trace_viewer/core/trace_model/process_memory_dump.html [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+/**
+ * @fileoverview Provides the ProcessMemoryDump class.
+ */
+tv.exportTo('tv.c.trace_model', function() {
+  /**
+   * The ProcessMemoryDump represents a memory dump of a single process.
+   * @constructor
+   */
+  function ProcessMemoryDump(start, worldMemoryDump) {
```
Comment (dj2) [inline trace_viewer/core/trace_model/process_memory_dump.html]: So, the process holds the world memory dumps, and not process level dumps? What's the relation between world and process dumps?

File: trace_viewer/core/trace_model/world_memory_dump.html [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+<script>
+'use strict';
+
+/**
+ * @fileoverview Provides the WorldMemoryDump class.
+ */
+tv.exportTo('tv.c.trace_model', function() {
+  /**
+   * The WorldMemoryDump represents a simultaneous memory dump of all processes.
```
Comment (dj2) [inline trace_viewer/core/trace_model/world_memory_dump.html]: This isn't related to V8 Isolate Worlds is it? If not, can we use something other then world? (Not that I have a good suggestion, just makes me think v8 isolated world when I read it)

Comment (petrcermak) [inline trace_viewer/core/trace_model/world_memory_dump.html]: OK. How about <tt>GlobalMemoryDump</tt>?

Comment (dj2) [inline trace_viewer/core/trace_model/world_memory_dump.html]: That sounds good. GlobalMemoryDumps (it seems like it holds many memory dumps?)

File: trace_viewer/core/trace_model/process_memory_dump.html (hunk 1 above)
Comment (petrcermak) [inline trace_viewer/core/trace_model/process_memory_dump.html]: A *single* <tt>WorldMemoryDump</tt> object links to *multiple* <tt>ProcessMemoryDump</tt> objects which were taken simultaneously. Every <tt>ProcessMemoryDump</tt> points to the <tt>WorldMemoryDump</tt> it belongs to (that's the <tt>worldMemoryDump</tt> argument of the constructor above). The actual memory data will be in the process memory dumps (once we know what it will look like).

File: trace_viewer/core/trace_model/world_memory_dump.html (hunk 2 above)
Comment (petrcermak) [inline trace_viewer/core/trace_model/world_memory_dump.html]: I think we should keep it singular. A single <tt>GlobalMemoryDump</tt> contains a <tt>ProcessMemoryDump</tt> for each process at a *single* time point.

Comment (dj2) [inline trace_viewer/core/trace_model/world_memory_dump.html]: Ah, that makes sense, thanks for explaining.

File: trace_viewer/core/trace_model/process_memory_dump.html (hunk 1 above)
Comment (dj2) [inline trace_viewer/core/trace_model/process_memory_dump.html]: So, why do we need to circular dependency? World -> Process -> World? Do processes need to look back up at their world parent for information?

Comment (petrcermak) [inline trace_viewer/core/trace_model/process_memory_dump.html]: We probably don't need it at the moment, but it is likely going to be useful later when we do visualisation (so that you could go from a single process view to the global view).

Comment (dj2) [inline trace_viewer/core/trace_model/process_memory_dump.html]: If we don't need it yet, let's leave it out. We can easily add it in if and when we do need it.

Comment (petrcermak) [inline trace_viewer/core/trace_model/process_memory_dump.html]: Fair point, I'll leave it out.

**Comment by petrcermak** [comment]
Here's an updated version after @natduca's and @dj2's comments. I made the following changes:

* renamed <tt>WorldMemoryDump</tt> to <tt>GlobalMemoryDump</tt> to avoid ambiguity with V8 Isolate Worlds,
* removed unnecessary methods (<tt>addMemoryDump</tt> etc.), and
* added a <tt>traceModelBounds_MemoryDumps</tt> *test* and augmented the <tt>traceModel_iterateAllEvents</tt> *test* to check that the relevant methods iterate over <tt>trace_model.globalMemoryDumps</tt> as well.

Regarding *importing* and *sorting* the dumps, I think that it should be all done by <tt>TraceEventImporter</tt> to avoid polluting <tt>TraceModel</tt> with temporary structures (namely the mapping from dump IDs to lists of process memory dumps). I plan to implement this functionality in a separate patch.

**Comment by natduca** [comment]
Sorting should be done in the model. The importing should be done in the
importers. The way to think of this is what would happen if there are N
importers... the model finalize is responsible for the integrity of the
model. The importers just do conversion.

**Comment by natduca** [comment]
And wrt the "dont have it now since its not used" this was done at my
suggestion. The thing is, this'll be trickier to add when someone needs it
later on and they don't know the structure of how they're created. That is,
having this circular data will be needed in the future (and we do it
everywhere in the model elsewhere) so I'd prefer we have it here from the
start.

Btw: make sure ProcessDump knows its global dump as well as its parent...

**Comment by petrcermak** [comment]
OK, that makes sense. Does that mean that we want the importer simply to build the dump ID → [<tt>ProcessMemoryDump</tt>] mapping and the sorting, building the <tt>GlobalMemoryDump</tt>, etc. done in the finalisation step in the <tt>TraceModel</tt>?

By "parent", do you mean the <tt>Process</tt> it belongs to? Isn't PID enough?

**Comment by natduca** [comment]
Build the dumps in the importer. Sort in the model.

I'd like to see direct links between objects, eg process, not pid.

**Comment by natduca** [comment]
(Rationale: the fact that dumps are stored as per process things plus ids
and need stitching is an artefact of how trace_event eg chrome are
generating the dumps. In some hypothetical other system, there might be a
different way that the dumps come in, so we want to support that too. The
model, however, really wants things sorted by timestamp so that it can,
say, find a nearby dump efficiently. So thats why that wants to be in the
model.)

Technically pid is enough. But, we have the process objects. So why add an
indirection? And, imagine down the road when you're trying to get someone
fresh to the codebase up to speed: "just grab the process" has to be said
as "go to the world dump then go from there to the model object then look
up this.pid in model.processes". Versus "just use this.process" :) I tend
to prefer choices that allow easy contributions by new team members,
basically.

Btw, the world dump has a pointer to the model right?

**Comment by petrcermak** [comment]
OK. I changed it so that:

* Only sorting is done in <tt>TraceModel</tt>. I added two tests for this.
* Building the dump objects (as well as building the <tt>GlobalMemoryDump</tt> ↔ <tt>ProcessMemoryDump</tt> relationship) is the responsibility of the *importer*. Hence, the functionality will be added in a subsequent patch.
* <tt>GlobalMemoryDump</tt> has a pointer to <tt>TraceModel</tt>.
* <tt>ProcessMemoryDump</tt> has a pointer to <tt>Process</tt>.

If you are now happy with the patch, I will land it.

**Comment by petrcermak** [comment]
I implemented all your comments :-)

**Comment by petrcermak** [comment]
I modified everything as you suggested except for the sorting tests because they are already using out-of-order input.

**Comment by natduca** [comment]
Sg. Merge away!
